kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/clusternetworkaddressconfigprovider/RangeAwarePortPerNodeClusterNetworkAddressConfigProvider.java

```diff
@@ -0,0 +1,257 @@
+ *    <li>{@code brokerStartPort} (optional) defines the starting range of port number that will be assigned to the brokers.  If omitted, it is defaulted to
+ *    the port number of {@code bootstrapAddress + 1}.</li>
+ *    <li>{@code nodeIdRanges} (required) defines the node id ranges present in the target cluster</li>
+ * </ul>
+ */
+public class RangeAwarePortPerNodeClusterNetworkAddressConfigProvider implements ClusterNetworkAddressConfigProvider {
+
+    private final HostPort bootstrapAddress;
+    private final String nodeAddressPattern;
+    private final int nodeStartPort;
+    private final Set<Integer> exclusivePorts;
+    private final Map<Integer, Integer> nodeIdToPortOffset;
+
+    /**
+     * Creates the provider.
+     *
+     * @param config configuration
+     */
+    public RangeAwarePortPerNodeClusterNetworkAddressConfigProvider(RangeAwarePortPerNodeClusterNetworkAddressConfigProviderConfig config) {
+        this.bootstrapAddress = config.bootstrapAddress;
+        this.nodeAddressPattern = config.nodeAddressPattern;
+        this.nodeStartPort = config.nodeStartPort;
+        this.nodeIdToPortOffset = config.nodeIdToPortOffset;
+
+        int nodeEndPortExclusive = nodeStartPort + config.nodeIdToPortOffset.size();
+        var exclusivePorts = IntStream.range(nodeStartPort, nodeEndPortExclusive).boxed().collect(Collectors.toCollection(HashSet::new));
+        exclusivePorts.add(bootstrapAddress.port());
+        this.exclusivePorts = Collections.unmodifiableSet(exclusivePorts);
+    }
+
+    @Override
+    public HostPort getClusterBootstrapAddress() {
+        return this.bootstrapAddress;
+    }
+
+    @Override
+    public HostPort getBrokerAddress(int nodeId) throws IllegalArgumentException {
+        if (!nodeIdToPortOffset.containsKey(nodeId)) {
+            throw new IllegalArgumentException(
+                    "Cannot generate node address for node id %d as it is not contained in the ranges defined for provider with downstream bootstrap %s"
+                            .formatted(
+                                    nodeId,
+                                    bootstrapAddress));
+        }
+        int port = nodeStartPort + nodeIdToPortOffset.get(nodeId);
+        return new HostPort(BrokerAddressPatternUtils.replaceLiteralNodeId(nodeAddressPattern, nodeId), port);
+    }
+
+    @Override
+    public Set<Integer> getExclusivePorts() {
+        return this.exclusivePorts;
+    }
+
+    @Override
+    public Map<Integer, HostPort> discoveryAddressMap() {
+        return nodeIdToPortOffset.keySet().stream()
+                .collect(Collectors.toMap(Function.identity(), this::getBrokerAddress));
+    }
+
+    /**
+     * @param startInclusive the (inclusive) initial value
+     * @param endExclusive the exclusive upper bound
+     */
+    public record IntRangeSpec(@JsonInclude(JsonInclude.Include.ALWAYS) @JsonProperty(required = true) int startInclusive,
```

Comment:
Why the need for the separate Spec object?  Why can't we just use Range?

Reply:
One reason is we can delay the validation so that if an invalid range is specified the error can be associated with a range name and present a more useful message identifying the problem range.